src/main/java/uk/gov/hmcts/reform/et/syaapi/service/AcasService.java

```diff
@@ -0,0 +1,125 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Pattern;
+
+/**
+ * This provides services to access the ACAS external service for retrieving ACAS Certificate's held in {@link
+ * AcasCertificate} objects.
+ */
+@Service
+public class AcasService {
+
+    public static final String OCP_APIM_SUBSCRIPTION_KEY = "Ocp-Apim-Subscription-Key";
+    public static final String VALID_ACAS_NUMBER_REGEX = "[a-zA-Z]{1,2}[\\d]{6}/[\\d]{2}/[\\d]{2}";
+    public static final int MAX_ACAS_RETRIES = 5;
+    private final RestTemplate restTemplate;
+    private final String acasApiUrl;
+    private final String acasApiKey;
+
+    /**
+     * Constructs an {@link AcasService} instance with the RestTemplate to use for talking with the ACAS service.
+     *
+     * @param restTemplate the RestTemplate to use for talking with the ACAS service
+     * @param acasApiUrl   the URL to access the ACAS API
+     * @param acasApiKey   the OCP APIM Subscription Key used in the header to authenticate when contacting ACAS
+     */
+    public AcasService(RestTemplate restTemplate,
+                       @Value("${acas.api.url}") String acasApiUrl,
+                       @Value("${acas.api.key}") String acasApiKey) {
+        this.restTemplate = restTemplate;
+        this.acasApiUrl = acasApiUrl;
+        this.acasApiKey = acasApiKey;
+    }
+
+    /**
+     * This will call upon ACAS with a set of ACAS case numbers to retrieve their associated certificates. Validation of
+     * the ACAS numbers is first applied and may result in an {@link InvalidAcasNumbersException} being thrown should
+     * there be any problems found. If all ACAS numbers are valid, then the service will attempt to retrieve a list of
+     * available {@link AcasCertificate}'s associated to the ACAS numbers provided. The service will retry up to 5 times
+     * to retrieve them if the call results in a 404 or 401 before then throwing an {@link AcasException} with the
+     * associated cause.
+     *
+     * @param acasNumbers are the ACAS numbers we are seeking Certificates for
+     * @return a List of {@link AcasCertificate}'s associated to the provided acasNumbers that are available at ACAS
+     * @throws AcasException               if a problem occurs obtaining the certificates.
+     * @throws InvalidAcasNumbersException if any of the acas numbers provided are invalid.
+     */
+    public List<AcasCertificate> getCertificates(String... acasNumbers)
+        throws InvalidAcasNumbersException, AcasException {
+
+        validateAcasNumbers(acasNumbers);
+        return attemptWithRetriesToFetchAcasCertificates(0, acasNumbers);
+    }
+
+    private List<AcasCertificate> attemptWithRetriesToFetchAcasCertificates(int attempts, String... acasNumbers)
+        throws AcasException {
+        try {
+            return fetchAcasCertificates(acasNumbers).getBody();
+        } catch (HttpClientErrorException e) {
```

Comment:
HttpClientErrorException is for 4xx responses (Bad request). Why retry on that error specifically rather than just RestClientException which is what RestTemplate.exchange throws?

Reply:
Totally agreed, thank James, good spot ;)